### ISSUE TITLE
Rounding errors

### ISSUE THREAD

**Issue by fjbender** [PAYONE-GmbH/prestashop-16#12]
Under certain circumstances, the cart and the amount a given request might be a cent or two off. The request will then be denied by the Payone platform. Example:

```
addressaddition=
aid=xxxx
amount=9823
api_version=3.10
birthday=00000000
city=Kiel
clearingsubtype=POV
clearingtype=rec
company=
country=DE
currency=EUR
customerid=20
de[1]=Faded Short Sleeve T-shirts
email=pui-test@payone.de
encoding=UTF-8
firstname=Paul
gender=f
id[1]=demo_1
integrator_name=prestashop
integrator_version=1.6.1.11
ip=10.0.4.39
it[1]=goods
key=xxxx
language=de
lastname=Payer
mid=xxxx
mode=test
narrative_text=BS Payone - gesicherter Rechnungskauf
no[1]=5
personalid=
portalid=xxxx
pr[1]=1965
reference=46bf017
request=authorization
salutation=Mrs.
shipping_city=Kiel
shipping_company=
shipping_country=DE
shipping_firstname=Paul
shipping_lastname=Payer
shipping_street=Zufallsweg 8
shipping_zip=24118
solution_name=fatchip
solution_version=1.2.1
street=Zufallsweg 8
telephonenumber=0431259681533
va[1]=1900
vatid=
zip=24118
```

This is probably due to the higher precision in Prestashop's cart/item logic. We can only work with two decimals while prestashop works with more. This should be avoided.

**Comment by paeddl** [comment]
This problem originates from the way how PrestaShop does rounding, and that PS always works with prices excluding tax.
In the PrestaShop backend, there are three different setup options for rounding.
Given the following cart (as in your example):
a product with 19,65€ incl. tax, 5 times in cart, shows the following total price in PrestaShop cart view:
- Round type "Round on each item": 98,25 € 
- Round type "Round on each line": 98,23 €
- Round type "Round on the total": 98,23 €

So basically, it is correct to send the 98,23€, because that is exactly what PrestaShop is calculating in this setup mode, and also what PrestaShop displays to the customers, renders in the invoice and stores in the database. Most merchants in Germany use the method "Round on the total". 

If we always send the "Round on each item"-total to payone, we have the problem that the amount at payone differs from the amount in PrestaShop.

**Comment by fjbender** [comment]
When switching to "Round on each item" everything seems to be correct. Maybe we should record that as a general suggestion for merchants.

**Comment by paeddl** [comment]
Unfortunately PrestaShop shows the following hint in its settings:

> Wählen Sie Rundungsmethode: Pro Artikel, pro Zeile oder Gesamtsumme runden (letztere für Deutschland empfohlen)

But yes, I think the best way would be to suggest that to merchants if they get back to you about this topic. But furthermore, I don't think we can do anything here at the plugin to avoid this behavior.

**Comment by fjbender** [comment]
Ok, thanks.
